File: socket_server.py
import socket
import json

def server_program(port=5000):
    # get the hostname
    host = socket.gethostname()

    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)  # get instance
    # look closely. The bind() function takes tuple as argument
    server_socket.bind((host, port))  # bind host address and port together

    # configure how many client the server can listen simultaneously
    server_socket.listen(2)
    conn, address = server_socket.accept()  # accept new connection
    print("Connection from: " + str(address))
    while True:
        info=""
        # receive data stream. it won't accept data packet greater than 1024 bytes
        time = conn.recv(1024).decode()
        data = [json.loads(line)
            for line in open('test_data.json', 'r', encoding='utf-8')]
        for line in data:
            if line["transactTime"] == str(time):
                info= str(line)
                break



        if not time:
            # if data is not received break
            break
        #print("from connected user: " + str(data))
        if info=="":
            info="no data or wrong data entered"
        conn.send(info.encode())  # send data to the client

    conn.close()  # close the connection


if __name__ == '__main__':
    port = int (input("enter port: "))
    server_program(port)

File: socket_server.cpp
// g++ socket_server.cpp -std=c++17 -o server && ./server

#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <map>
#include <arpa/inet.h>
#include <unistd.h>
#include "nlohmann/json.hpp"

#define SERVER_IP "127.0.0.1"
#define PORT 5000
#define JSON_FILE "test_data.json"

using json = nlohmann::json;

void server_program(int port)
{
    std::string host = SERVER_IP;

    int server_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (server_socket == -1)
    {
        perror("Error creating socket");
        exit(1);
    }

    sockaddr_in server_addr{};
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(port);
    server_addr.sin_addr.s_addr = INADDR_ANY;

    if (bind(server_socket, reinterpret_cast<struct sockaddr *>(&server_addr), sizeof(server_addr)) < 0)
    {
        perror("Bind failed");
        exit(1);
    }

    if (listen(server_socket, 2) < 0)
    {
        perror("Listen failed");
        exit(1);
    }

    std::cout << "Server listening on port " << port << std::endl;

    sockaddr_in client_addr{};
    socklen_t client_addr_len = sizeof(client_addr);
    int client_socket = accept(server_socket, reinterpret_cast<struct sockaddr *>(&client_addr), &client_addr_len);
    if (client_socket < 0)
    {
        perror("Accept failed");
        exit(1);
    }

    std::cout << "Client connected" << std::endl;

    char buffer[1024] = {0};
    while (true)
    {
        std::string time;
        memset(buffer, 0, sizeof(buffer));
        if (recv(client_socket, buffer, sizeof(buffer) - 1, 0) == -1)
        {
            perror("Error receiving data");
            exit(1);
        }
        time = buffer;

        if (time.empty() || time == "bye")
            break;

        std::ifstream file(JSON_FILE);
        if (!file)
        {
            std::cerr << "Error opening file: test_data.json" << std::endl;
            break;
        }

        std::string line;
        std::string info = "No data or wrong data entered";
        while (std::getline(file, line))
        {
            try
            {
                json data = json::parse(line);
                std::cout << "json line : " << data << std::endl;

                if (data["transactTime"] == time)
                {
                    info = data.dump();
                    break;
                }
            }
            catch (const json::parse_error &e)
            {
                std::cerr << "Error parsing JSON: " << e.what() << std::endl;
            }
        }
        file.close();

        if (send(client_socket, info.c_str(), info.size(), 0) == -1)
        {
            perror("Error sending data");
            exit(1);
        }
    }

    close(client_socket);
    close(server_socket);
}

int main()
{
    server_program(PORT);
    return 0;
}
